fdm-core/src/authorization.test.ts

```diff
@@ -0,0 +1,366 @@
+                .select()
+                .from(authZSchema.role)
+                .where(
+                    and(
+                        eq(authZSchema.role.resource, "farm"),
+                        eq(authZSchema.role.resource_id, farm_id),
+                        eq(authZSchema.role.principal_id, principal_id),
+                        eq(authZSchema.role.role, "owner"),
+                        isNotNull(authZSchema.role.deleted),
+                    ),
+                )
+            expect(roles.length).toBe(1)
+        })
+
+        it("should not throw an error when revoking a non-existing role", async () => {
+            await revokeRole(fdm, "farm", "owner", farm_id, principal_id)
+            const roles = await fdm
+                .select()
+                .from(authZSchema.role)
+                .where(
+                    and(
+                        eq(authZSchema.role.resource, "farm"),
+                        eq(authZSchema.role.resource_id, farm_id),
+                        eq(authZSchema.role.principal_id, principal_id),
+                        eq(authZSchema.role.role, "owner"),
+                        isNotNull(authZSchema.role.deleted),
+                    ),
+                )
+            expect(roles.length).toBe(0)
+        })
+
+        it("should throw an error for invalid resource", async () => {
+            await expect(
+                revokeRole(
+                    fdm,
+                    // biome-ignore lint/suspicious/noExplicitAny: Used for testing validation
+                    "unknown_resource" as any,
+                    "owner",
+                    farm_id,
+                    principal_id,
+                ),
+            ).rejects.toThrowError()
+        })
+
+        it("should throw an error for invalid role", async () => {
+            await expect(
+                revokeRole(
+                    fdm,
+                    "farm",
+                    // biome-ignore lint/suspicious/noExplicitAny: Used for testing validation
+                    "unknown_role" as any,
+                    farm_id,
+                    principal_id,
+                ),
+            ).rejects.toThrowError()
+        })
+    })
+
+    describe("listResources", () => {
+        it("should list resources a principal has access to", async () => {
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id)
+
+            const accessibleResources = await listResources(
+                fdm,
+                "farm",
+                "read",
+                principal_id,
+            )
+            expect(accessibleResources).toContain(farm_id)
+        })
+
+        it("should handle multiple roles", async () => {
+            const principal_id_new = createId()
+            const farm_id2 = createId()
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id_new)
+            await grantRole(fdm, "farm", "advisor", farm_id2, principal_id_new)
+
```

Comment:
different farm id for advisor than for owner?

Reply:
Yes, user should be able to be owner of farm 1 and be advisor of farm 2

Reply:
agreed. max. flexibility ;-)